File: run-bugs.py
#!/usr/bin/env python3
"""Query models on 50 bugs with realistic prompts.

- blind: stack trace + bug description, no code = pure guesswork
- ctx: stack trace + the buggy function body (simulating `ctx` tool output)

Usage:
  python3 run-bugs.py [--count 50] [--cats 3B-blind,3B-ctx]
"""
import json, os, sys, random, argparse, textwrap

BUGS_FILE = os.path.join(os.path.dirname(__file__), 'bugs.jsonl')
OUT = os.path.join(os.path.dirname(__file__), 'bugfix')
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '.venv', 'lib', 'python3.12', 'site-packages'))

MODEL_PATH_3B = os.path.join(os.path.dirname(__file__), 'data/models/qwen2.5-coder-3b-instruct.gguf')
MODEL_PATH_7B = os.path.join(os.path.dirname(__file__), 'data/models/qwen2.5-coder-7b-instruct.gguf')
API_KEY = os.environ.get('OPENROUTER_API_KEY')
MODEL_32B = 'qwen/qwen-2.5-coder-32b-instruct'

SYSTEM_PROMPT = 'You are a senior Ruby developer. Fix the bug in the code shown below. Return ONLY the corrected Ruby code in a ```ruby block.'

def repo_path_for(bug):
    repo = bug.get('repo') or {}
    return repo.get('repo_path') or os.path.join(os.path.dirname(__file__), '.eval', 'cheat')

def file_rel_for(bug):
    return bug.get('file_rel') or bug['file'].replace('/home/yahn/cheat/', '')

def source_path_for(bug):
    return os.path.join(repo_path_for(bug), file_rel_for(bug))

def simulated_worktree_state(bug):
    discovery = bug.get('discovery') or {
        'scenario': 'production_stack_trace',
        'worktree_dirty': False,
        'dirty_scope': 'clean_tree',
        'dirty_files': [],
        'hint': 'Clean tree; treat this as a production/existing bug that slipped through tests.'
    }
    file_rel = file_rel_for(bug)
    dirty_files = discovery.get('dirty_files') or []
    target_dirty = [f for f in dirty_files if f.get('file_rel') == file_rel]
    lines = [
        '=' * 60,
        'WORKTREE STATE',
        '=' * 60,
        f"Scenario: {discovery.get('scenario', 'unknown')}",
        f"Worktree: {'dirty' if discovery.get('worktree_dirty') else 'clean'}",
    ]
    if target_dirty:
        lines.append(f"Target file: dirty ({file_rel})")
        for entry in target_dirty:
            for r in entry.get('line_ranges', []):
                lines.append(f"  lines {r.get('start')}-{r.get('end')}: {r.get('reason')}")
    else:
        lines.append(f"Target file: clean ({file_rel})")
    other_dirty = [f for f in dirty_files if f.get('file_rel') != file_rel]
    if other_dirty:
        lines.append('Other dirty files:')
        for entry in other_dirty[:10]:
            lines.append(f"  {entry.get('status', 'modified')} {entry.get('file_rel')} ({entry.get('role', 'unknown')})")
    lines.append(f"Interpretation: {discovery.get('hint', 'unknown')}")
    return '\n'.join(lines)

def prompt_blind(bug):
    file_rel = file_rel_for(bug)
    func = bug['function'].split('.')[-1]
    return textwrap.dedent(f"""\
File: {file_rel}
Function: {func}

{bug['prompt']}

Return ONLY the corrected function `{func}` in a ```ruby block.""")

def prompt_with_context(bug):
    file_rel = file_rel_for(bug)
    func = bug['function'].split('.')[-1]
    # Simulate what `ctx {file}#{line}` would return: the buggy function body
    # (We show the mutated body, which is what ctx would see on disk after the bug was introduced)
    return textwrap.dedent(f"""\
File: {file_rel}
Function: {func}

Here is the current (buggy) code of function `{func}`:

```ruby
{build_function_with_context(bug)}
```

{simulated_worktree_state(bug)}

{bug['prompt']}

Return ONLY the corrected function `{func}` in a ```ruby block.""")

def build_function_with_context(bug):
    """Reconstruct the buggy function with its def signature and mutated body."""
    filepath = source_path_for(bug)
    mutated = bug['mutated_body']

    try:
        with open(filepath) as f:
            source = f.read().replace('\r\n', '\n')
        source = source.replace(bug['original_body'].replace('\r\n', '\n'), mutated.replace('\r\n', '\n'), 1)
        lines = source.splitlines()
        start = bug.get('function_start_line')
        end = bug.get('function_end_line')
        if start and end:
            return '\n'.join(lines[start - 1:end])
    except:
        pass
    # Fallback: just show the mutated body
    return mutated

def load_model(path):
    from llama_cpp import Llama
    print(f"Loading: {path}")
    return Llama(model_path=path, n_ctx=4096, n_threads=32, verbose=False, n_gpu_layers=0)

def query_gguf(llm, prompt):
    full = '<|im_start|>system\n' + SYSTEM_PROMPT + '\n<|im_end|>\n<|im_start|>user\n' + prompt + '\n<|im_end|>\n<|im_start|>assistant\n```ruby\n'
    output = llm(full, max_tokens=1024, temperature=0.1, stop=['<|im_end|>', '<|end|>', '\n```'])
    return '```ruby\n' + output['choices'][0]['text'].strip() + '\n```'

def query_api(prompt):
    import requests
    if not API_KEY:
        return '[[SKIPPED: API key not set]]'
    resp = requests.post(
        'https://openrouter.ai/api/v1/chat/completions',
        headers={'Authorization': f'Bearer {API_KEY}', 'Content-Type': 'application/json'},
        json={'model': MODEL_32B, 'messages': [
            {'role': 'system', 'content': SYSTEM_PROMPT},
            {'role': 'user', 'content': prompt},
        ], 'max_tokens': 1500, 'temperature': 0.1},
        timeout=120,
    ).json()
    if 'error' in resp:
        raise Exception(f"API error: {resp['error']}")
    return resp['choices'][0]['message']['content'].strip()

# === Main ===
parser = argparse.ArgumentParser()
parser.add_argument('--count', type=int, default=50)
parser.add_argument('--cats', type=str, default='')
parser.add_argument('--dry-run-prompts', action='store_true', help='write .prompt.txt files and skip model calls')
args = parser.parse_args()

with open(BUGS_FILE) as f:
    bugs = [json.loads(line) for line in f if line.strip()]
random.seed(42)
sample = random.sample(bugs, min(args.count, len(bugs)))
n = len(sample)
print(f"Sample: {n} bugs\n")

cats = args.cats.split(',') if args.cats else ['3B-blind', '3B-ctx', '7B-blind', '32B-blind', '32B-ctx']
blinds = [prompt_blind(b) for b in sample]
ctxs = [prompt_with_context(b) for b in sample]

models = {}

for cat in cats:
    dirpath = os.path.join(OUT, cat)
    os.makedirs(dirpath, exist_ok=True)
    prompts = ctxs if 'ctx' in cat else blinds

    for i, (bug, prompt) in enumerate(zip(sample, prompts)):
        fpath = os.path.join(dirpath, f'{i+1:02d}.txt')
        ppath = os.path.join(dirpath, f'{i+1:02d}.prompt.txt')
        label = f'[{cat}] bug {i+1}/{n}'
        print(f'  {label}', flush=True)
        with open(ppath, 'w') as f:
            f.write(prompt + '\n')
        if args.dry_run_prompts:
            continue

        try:
            if cat.startswith('3B'):
                if '3B' not in models:
                    models['3B'] = load_model(MODEL_PATH_3B)
                result = query_gguf(models['3B'], prompt)
            elif cat.startswith('7B'):
                if '7B' not in models:
                    models['7B'] = load_model(MODEL_PATH_7B)
                result = query_gguf(models['7B'], prompt)
            elif cat.startswith('32B'):
                result = query_api(prompt)
            else:
                result = '[[UNKNOWN CATEGORY]]'

            with open(fpath, 'w') as f:
                f.write(result + '\n')
        except Exception as e:
            with open(fpath, 'w') as f:
                f.write(f'[[ERROR: {e}]]\n')

print('\nDone.')
if args.dry_run_prompts:
    print('Dry run: wrote prompt files only.')

File: src/run-bugs.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# run-bugs.rb — Query models on bugs.
#
# - blind: stack trace + bug description only
# - ctx: stack trace + buggy function body (simulating `ctx` tool)
#
# Usage:
#   ruby src/run-bugs.rb [--count 50] [--cats 3B-blind,3B-ctx,...]
#   OPENROUTER_API_KEY=sk-... ruby src/run-bugs.rb --cats 32B-blind,32B-ctx

require "optparse"
require "json"
require "fileutils"
require "net/http"
require "uri"
require "open3"
require "tempfile"

ROOT = File.expand_path('..', __dir__)
BUGS_FILE = File.join(ROOT, "bugs.jsonl")
OUT       = File.join(ROOT, "bugfix")
MODEL_3B  = File.join(ROOT, "data/models/qwen2.5-coder-3b-instruct.gguf")
MODEL_7B  = File.join(ROOT, "data/models/qwen2.5-coder-7b-instruct.gguf")
VENV_PY   = File.join(ROOT, ".venv/bin/python3")
API_KEY   = ENV["OPENROUTER_API_KEY"]
MODEL_32B = "qwen/qwen-2.5-coder-32b-instruct"

SYS = "You are a senior Ruby developer. Fix the bug. Return ONLY corrected code in a ```ruby block."

opts = { count: 50, cats: "3B-blind,3B-ctx,7B-blind,32B-blind,32B-ctx" }
OptionParser.new do |o|
  o.banner = "Usage: ruby src/run-bugs.rb [options]"
  o.on("--count N", Integer, "Number of bugs (default: 50)") { |v| opts[:count] = v }
  o.on("--cats LIST", "Comma-separated categories (default: all 5)") { |v| opts[:cats] = v }
  o.on("-h", "--help") { puts o; exit 0 }
end.parse!

cats = opts[:cats].split(",").map(&:strip)
count = opts[:count]

# ── Prompts ────────────────────────────────────────────

def def_line(filepath, func)
  File.readlines(filepath).find { |l| l =~ /def\s+(self\.)?#{Regexp.escape(func)}\b/ }&.rstrip || "def #{func}"
end

def prompt_blind(bug)
  rel  = bug["file"].sub(%r{^/home/yahn/cheat/}, "")
  func = bug["function"].split(".").last
  src  = File.read(bug["file"])
  <<~PROMPT
    File: #{rel}
    Function: #{func}

    Full source file:
    ```ruby
    #{src}
    ```

    Error:
    #{bug["stack_trace"]}
    #{bug["prompt"]}
    Return ONLY the corrected function `#{func}` in a ```ruby block.
  PROMPT
end

def prompt_with_context(bug)
  rel   = bug["file"].sub(%r{^/home/yahn/cheat/}, "")
  func  = bug["function"].split(".").last
  src   = File.read(bug["file"])
  dline = def_line(bug["file"], func)
  <<~PROMPT
    File: #{rel}
    Function: #{func}

    Full source file:
    ```ruby
    #{src}
    ```

    The buggy function (#{func}):
    ```ruby
    #{dline}
    #{bug["mutated_body"]}
    ```

    Error:
    #{bug["stack_trace"]}
    #{bug["prompt"]}
    Return ONLY the corrected function `#{func}` in a ```ruby block.
  PROMPT
end

# ── GGUF via llama_cpp (Python subprocess) ─────────────

def query_gguf(model_path, prompt)
  tmp = Tempfile.new(["llama_prompt", ".py"])
  tmp.write(<<~PYTHON)
    import sys, json, tempfile, os
    from llama_cpp import Llama
    prompt_file = #{tmp.path.sub(/\.py\z/, '.txt').to_json}
    with open(prompt_file) as f:
        user_msg = f.read()
    llm = Llama(model_path=#{model_path.to_json}, n_ctx=32768, n_threads=32, verbose=False)
    sys_msg = #{SYS.to_json}
    p = f"<|im_start|>system\\n{sys_msg}\\n<|im_end|>\\n<|im_start|>user\\n{user_msg}\\n<|im_end|>\\n<|im_start|>assistant\\n```ruby\\n"
    r = llm(p, max_tokens=1024, temperature=0.1, stop=["<|im_end|>", "<|end|>"])
    print(r["choices"][0]["text"].strip())
  PYTHON
  tmp.close
  prompt_file = tmp.path.sub(/\.py\z/, ".txt")
  File.write(prompt_file, prompt)
  out, err, st = Open3.capture3(VENV_PY, tmp.path)
  File.unlink(tmp.path, prompt_file) rescue nil
  raise err unless st.success?
  "```ruby" + out.strip
rescue => e
  "[[GGUF_ERROR: #{e.message[0..200]}]]"
end

# ── OpenRouter ─────────────────────────────────────────

def query_openrouter(prompt)
  return "[[SKIPPED: OPENROUTER_API_KEY not set]]" unless ENV["OPENROUTER_API_KEY"]
  uri  = URI("https://openrouter.ai/api/v1/chat/completions")
  body = {
    model:    MODEL_32B,
    messages: [
      { role: "system", content: SYS },
      { role: "user",   content: prompt },
    ],
    max_tokens: 1500, temperature: 0.1,
  }.to_json
  req = Net::HTTP::Post.new(uri,
    "Content-Type"  => "application/json",
    "Authorization" => "Bearer #{ENV["OPENROUTER_API_KEY"]}")
  resp = Net::HTTP.start(uri.hostname, uri.port, use_ssl: true) { |http| http.request(req) }
  data = JSON.parse(resp.body)
  if data["error"]
    "[[API_ERROR: #{data["error"]["message"]}]]"
  else
    data.dig("choices", 0, "message", "content")&.strip || "[[EMPTY]]"
  end
rescue => e
  "[[API_ERROR: #{e.message}]]"
end

# ── Main ───────────────────────────────────────────────

sample = File.readlines(BUGS_FILE).map { |l| JSON.parse(l.strip) rescue nil }.compact
srand(42)
sample = sample.sample(count)
puts "Running #{sample.size} bugs\n"

blinds = sample.map { |b| prompt_blind(b) }
ctxs   = sample.map { |b| prompt_with_context(b) }

cats.each do |cat|
  dir = File.join(OUT, cat)
  FileUtils.mkdir_p(dir)
  prompts = cat.include?("ctx") ? ctxs : blinds

  sample.each_with_index do |_bug, i|
    fpath = File.join(dir, format("%02d.txt", i + 1))
    label = "[#{cat}] #{i + 1}/#{sample.size}"
    print "  #{label.ljust(25)}"
    $stdout.flush

    result = case cat
             when /\A3B/  then query_gguf(MODEL_3B, prompts[i])
             when /\A7B/  then query_gguf(MODEL_7B, prompts[i])
             when /\A32B/ then query_openrouter(prompts[i])
             else "[[UNKNOWN CATEGORY]]"
             end
    File.write(fpath, result + "\n")
    File.write(fpath.sub(/\.txt\z/, '.prompt.txt'), prompts[i] + "\n")
    puts "OK"
  end
end

puts "\nDone."
